File: studio-sdk/javascript/src/services/pollingService.js
export class PollingService extends EventTarget {
  #intervalId = null

  constructor(mediaId, apiService, { intervalMs = 1000 } = {}) {
    super()
    this.mediaId = mediaId
    this.apiService = apiService

    this.#intervalId = setInterval(async () => {
      try {
        const media = await this.apiService.getMediaStatus({ mediaId })
        const job = media?.jobs?.transcription

        switch (job?.state) {
          case "done": {
            const fullMedia = await this.apiService.getMedia({ mediaId })
            this.#clearInterval()
            this.dispatchEvent(new CustomEvent("done", { detail: fullMedia }))
            break
          }
          case "error": {
            this.#clearInterval()
            this.dispatchEvent(new CustomEvent("error", { detail: job }))
            break
          }
          default:
            this.dispatchEvent(new CustomEvent("update", { detail: job }))
            break
        }
      } catch (err) {
        this.#clearInterval()
        this.dispatchEvent(new CustomEvent("error", { detail: err }))
      }
    }, intervalMs)
  }

  stop() {
    if (this.#intervalId === null) {
      return
    }
    this.#clearInterval()
    this.dispatchEvent(new CustomEvent("cancelled"))
  }

  #clearInterval() {
    if (this.#intervalId !== null) {
      clearInterval(this.#intervalId)
      this.#intervalId = null
    }
  }
}

File: studio-sdk/python/linto/services/pollingService.py
import asyncio
import logging


class EventEmitter:
    def __init__(self):
        self._listeners = {}

    def on(self, event, callback):
        self._listeners.setdefault(event, []).append(callback)

    def off(self, event, callback):
        if event in self._listeners:
            self._listeners[event].remove(callback)

    async def emit(self, event, *args, **kwargs):
        for callback in self._listeners.get(event, []):
            result = callback(*args, **kwargs)
            if asyncio.iscoroutine(result):
                await result


class PollingService(EventEmitter):
    def __init__(self, media_id, api_service, interval=1.0):
        super().__init__()
        self.media_id = media_id
        self.api_service = api_service
        self.interval = interval
        self._task = None
        self.start()

    def start(self):
        if self._task is None:
            self._task = asyncio.create_task(self._poll_loop())

    def stop(self):
        if self._task:
            self._task.cancel()
            self._task = None

    async def _poll_loop(self):
        try:
            while True:
                media = await self.api_service.get_media_status(mediaId=self.media_id)
                job = media.get("jobs", {}).get("transcription")

                state = job.get("state") if job else None
                if state == "done":
                    logging.debug("Sending done event")
                    media = await self.api_service.get_media(mediaId=self.media_id)
                    await self.emit("done", media)
                    self.stop()
                    break
                elif state == "error":
                    logging.debug("Sending error event")
                    await self.emit("error")
                    self.stop()
                    break
                else:
                    logging.debug("Sending update event")
                    await self.emit("update", job)
                await asyncio.sleep(self.interval)
        except asyncio.CancelledError:
            logging.debug("job cancelled")
            pass
